File: server/tools/physics.go
package tools

import (
	"fmt"
	"math"
)

type Vector2 struct {
	X float64
	Y float64
}

func (v Vector2) Add(u Vector2) Vector2      { return Vector2{v.X + u.X, v.Y + u.Y} }
func (v Vector2) Subtract(u Vector2) Vector2 { return Vector2{v.X - u.X, v.Y - u.Y} }
func (v Vector2) Multiply(s float64) Vector2 { return Vector2{v.X * s, v.Y * s} }
func (v Vector2) Dot(u Vector2) float64      { return v.X*u.X + v.Y*u.Y }
func (v Vector2) LengthSquared() float64     { return v.Dot(v) }
func (v Vector2) Length() float64            { return math.Sqrt(v.LengthSquared()) }
func (v Vector2) Norm() Vector2 {
	if l := v.Length(); l != 0 {
		return v.Multiply(1 / l)
	}
	return Vector2{}
}

type Vector2Int struct {
	X int
	Y int
}

func (v Vector2Int) Add(u Vector2Int) Vector2Int      { return Vector2Int{v.X + u.X, v.Y + u.Y} }
func (v Vector2Int) Subtract(u Vector2Int) Vector2Int { return Vector2Int{v.X - u.X, v.Y - u.Y} }
func (v Vector2Int) DistanceSquared(u Vector2Int) int {
	return (v.X-u.X)*(v.X-u.X) + (v.Y-u.Y)*(v.Y-u.Y)
}

type WallCollisionType int

const (
	NONE WallCollisionType = iota
	BOTTOM
	TOP
	LEFT
	RIGHT
)

type Rect struct {
	topleft Vector2
	width   float64
	height  float64
}

type Circle struct {
	Center   Vector2
	Radius   float64
	Velocity Vector2
}

func (c *Circle) DistanceSquared(c1 *Circle) float64 {
	return c.Center.Subtract(c1.Center).LengthSquared()
}

type ShotData struct {
	direction Vector2
	power     int //should be a value from one to 5, could make this an enum.
}

const (
	dt         = 1.0 / 120.0
	drag       = 0.989
	stop2      = 1e-4
	settleNeed = 5
	maxSteps   = 30 * 120
)

func PhysicsResolver(activePlayer *Circle, playerPositions []*Circle, walls []*Rect, shotData ShotData) {
	ApplyImpulse(activePlayer, shotData)
	slowFrames := 0
	for step := 0; step < maxSteps; step++ {
		Integrate(playerPositions)
		ResolveCircleWallCollisions(playerPositions, walls)
		ResolveCircleCircleCollisions(playerPositions)
		ApplyFriction(playerPositions)

		if AllStopped(playerPositions) {
			slowFrames++
			if slowFrames >= settleNeed {
				ResetVelocities(playerPositions)
				break
			}
		} else {
			slowFrames = 0
		}
	}
}

// HELPER FUNCTIONS

func ApplyImpulse(circle *Circle, shotData ShotData) {
	circle.Velocity = shotData.direction.Norm().Multiply(float64(shotData.power))
}

func Integrate(circles []*Circle) {
	for i := range circles {
		circles[i].Center = circles[i].Center.Add(circles[i].Velocity.Multiply(dt))
	}
}

func CheckCircleCircleCollision(c1 Circle, c2 Circle) bool {
	return (c1.Center.Subtract(c2.Center).LengthSquared() < (c1.Radius+c2.Radius)*(c1.Radius+c2.Radius))
}

func CheckCircleWallCollision(c1 Circle, wall *Rect) WallCollisionType {
	threshold := -0.05

	bottomCheck := (c1.Center.Y + c1.Radius) - (wall.topleft.Y - wall.height)
	topCheck := (wall.topleft.Y) - (c1.Center.Y - c1.Radius)
	leftCheck := (c1.Center.X + c1.Radius) - (wall.topleft.X)
	rightCheck := (wall.topleft.X + wall.width) - (c1.Center.X - c1.Radius)

	if bottomCheck < threshold || topCheck < threshold || leftCheck < threshold || rightCheck < threshold {
		return NONE
	} else if bottomCheck > threshold && bottomCheck < -threshold {
		return BOTTOM
	} else if topCheck > threshold && topCheck < -threshold {
		return TOP
	} else if leftCheck > threshold && leftCheck < -threshold {
		return LEFT
	} else if rightCheck > threshold && rightCheck < -threshold {
		return RIGHT
	} else {
		fmt.Println("ERROR: Circle-Wall Collision failed -- CheckCircleWallCollision()")
		return NONE
	}
}

func DoPositionalCorrection(c1 *Circle, c2 *Circle) {
	delta := c1.Center.Subtract(c2.Center)
	dist := delta.Length()
	radSum := c1.Radius + c2.Radius

	penetration := radSum - dist
	normal := delta.Multiply(1 / dist)

	c1.Center = c1.Center.Add(normal.Multiply(penetration * 0.5))
	c2.Center = c2.Center.Add(normal.Multiply(penetration * 0.5))
}

func FixOverlapX(dir WallCollisionType, circle *Circle, wall *Rect) {
	if dir == LEFT {
		overlap := (circle.Center.X + circle.Radius) - wall.topleft.X
		if overlap > 0 {
			circle.Center.X -= (overlap + 0.1)
		}
	} else if dir == RIGHT {
		overlap := (circle.Center.X - circle.Radius) - (wall.topleft.X + wall.width)
		if overlap < 0 {
			circle.Center.X -= (overlap - 0.1)
		}
	}
}

func FixOverlapY(dir WallCollisionType, circle *Circle, wall *Rect) {
	if dir == TOP {
		overlap := (circle.Center.Y + circle.Radius) - wall.topleft.Y
		if overlap > 0 {
			circle.Center.Y -= (overlap + 0.1)
		}
	} else if dir == BOTTOM {
		overlap := (circle.Center.Y - circle.Radius) - (wall.topleft.Y + wall.height)
		if overlap < 0 {
			circle.Center.Y -= (overlap - 0.1)
		}
	}
}

func ResolveCircleCircleCollisions(circles []*Circle) {
	//for all pairs, check collision and resolve if colliding
	for i := 0; i < len(circles); i++ {
		for j := i + 1; j < len(circles); j++ {
			if CheckCircleCircleCollision(*circles[i], *circles[j]) {
				DoPositionalCorrection(circles[i], circles[j])
				relVel := circles[i].Velocity.Subtract(circles[j].Velocity)
				delta := circles[i].Center.Subtract(circles[j].Center)
				dist := delta.Length()
				normal := delta.Multiply(1 / dist)
				speed := relVel.Dot(normal)
				if speed >= 0 {
					continue
				}

				impulse := normal.Multiply(-speed)
				circles[i].Velocity = circles[i].Velocity.Add(impulse)
				circles[j].Velocity = circles[j].Velocity.Subtract(impulse)
			}
		}
	}
}

func ResolveCircleWallCollisions(circles []*Circle, walls []*Rect) {
	for i := range circles {
		for j := range walls {
			collisionDirection := CheckCircleWallCollision(*circles[i], walls[j])
			collisionDetected := false
			switch collisionDirection {
			case NONE:

			case TOP, BOTTOM:
				circles[i].Velocity.Y *= -1
				FixOverlapY(collisionDirection, circles[i], walls[j])
				collisionDetected = true
			case LEFT, RIGHT:
				circles[i].Velocity.X *= -1
				FixOverlapX(collisionDirection, circles[i], walls[j])
				collisionDetected = true
			}
			if collisionDetected {
				break
			}
		}
	}
}

func ApplyFriction(circles []*Circle) {
	for i := range circles {
		circles[i].Velocity = circles[i].Velocity.Multiply(drag)
	}
}

func AllStopped(circles []*Circle) bool {
	for i := range circles {
		if circles[i].Velocity.LengthSquared() >= stop2 {
			return false
		}
	}

	return true
}

func ResetVelocities(circles []*Circle) {
	for i := range circles {
		circles[i].Velocity = Vector2{0, 0}
	}
}

File: client/src/physics.ts
// Physics types and constants

import { PositionComponent } from "./Components";

export class Vector2 {
  constructor(public x: number, public y: number) {}

  add(v: Vector2): Vector2 {
    return new Vector2(this.x + v.x, this.y + v.y);
  }

  subtract(v: Vector2): Vector2 {
    return new Vector2(this.x - v.x, this.y - v.y);
  }

  multiply(s: number): Vector2 {
    return new Vector2(this.x * s, this.y * s);
  }

  dot(v: Vector2): number {
    return this.x * v.x + this.y * v.y;
  }

  lengthSquared(): number {
    return this.dot(this);
  }

  length(): number {
    return Math.sqrt(this.lengthSquared());
  }

  norm(): Vector2 {
    const len = this.length();
    return len !== 0 ? this.multiply(1 / len) : new Vector2(0, 0);
  }
}

export class Circle {
  velocity = new Vector2(0, 0);

  constructor(
    public center: Vector2,
    public radius: number
  ) {}

  update(cenX: number, cenY: number, velX: number, velY: number) {
    this.center.x = cenX;
    this.center.y = cenY;
    this.velocity.x = velX;
    this.velocity.y = velY;
  }
}

export class Rect {
  constructor(
    public topLeft: Vector2,
    public width: number,
    public height: number
  ) {}
}

export class ShotData {
  constructor(
    public direction: Vector2,
    public power: number
  ) {}
}

// Constants
const dt = 1 / 120;
const drag = 0.989;
const stop2 = 1e-4;
const settleNeed = 5;
const maxSteps = 30 * 120;

export interface PhysicsState {
  activeCircle: Circle;
  circles: Circle[];
  shot: ShotData;
  walls: Rect[];
  step: number;
  slowFrames: number;
  onComplete: () => void;
}

// Physics functions
function applyImpulse(circle: Circle, shot: ShotData): void {
  circle.velocity = shot.direction.norm().multiply(shot.power);
}

function integrate(circles: Circle[]): void {
  for (const c of circles) {
    c.center = c.center.add(c.velocity.multiply(dt));
  }
}

function checkCircleCircle(c1: Circle, c2: Circle): boolean {
  return c1.center.subtract(c2.center).lengthSquared() < (c1.radius + c2.radius) ** 2;
}

function doPositionalCorrection(c1: Circle, c2: Circle): void {
  const delta = c1.center.subtract(c2.center);
  const dist = delta.length();
  const radSum = c1.radius + c2.radius;
  const penetration = radSum - dist;
  if (dist === 0) return;

  const normal = delta.multiply(1 / dist);
  const correction = normal.multiply(penetration * 0.5);

  c1.center = c1.center.add(correction);
  c2.center = c2.center.subtract(correction);
}

function resolveCircleCircle(circles: Circle[]): void {
  for (let i = 0; i < circles.length; i++) {
    for (let j = i + 1; j < circles.length; j++) {
      const c1 = circles[i];
      const c2 = circles[j];
      if (checkCircleCircle(c1, c2)) {
        doPositionalCorrection(c1, c2);

        const relVel = c1.velocity.subtract(c2.velocity);
        const delta = c1.center.subtract(c2.center);
        const dist = delta.length();
        if (dist === 0) continue;

        const normal = delta.multiply(1 / dist);
        const speed = relVel.dot(normal);
        if (speed >= 0) continue;

        const impulse = normal.multiply(-speed);
        c1.velocity = c1.velocity.add(impulse);
        c2.velocity = c2.velocity.subtract(impulse);
      }
    }
  }
}

function resolveCircleWallCollisions(circles: Circle[], walls: Rect[]): void {
  for (const circle of circles) {
    for (const wall of walls) {
      const left = wall.topLeft.x;
      const right = wall.topLeft.x + wall.width;
      const top = wall.topLeft.y;
      const bottom = wall.topLeft.y + wall.height;

      // Check left/right collision
      if (circle.center.x - circle.radius < left) {
        circle.center.x = left + circle.radius;
        circle.velocity.x *= -1;
      } else if (circle.center.x + circle.radius > right) {
        circle.center.x = right - circle.radius;
        circle.velocity.x *= -1;
      }

      // Check top/bottom collision
      if (circle.center.y - circle.radius < top) {
        circle.center.y = top + circle.radius;
        circle.velocity.y *= -1;
      } else if (circle.center.y + circle.radius > bottom) {
        circle.center.y = bottom - circle.radius;
        circle.velocity.y *= -1;
      }
    }
  }
}

function applyFriction(circles: Circle[]): void {
  for (const c of circles) {
    c.velocity = c.velocity.multiply(drag);
  }
}

function allStopped(circles: Circle[]): boolean {
  return circles.every(c => c.velocity.lengthSquared() < stop2);
}

function resetVelocities(circles: Circle[]): void {
  for (const c of circles) {
    c.velocity = new Vector2(0, 0);
  }
}

// Main physics resolver
export function startPhysicsSimulation(
  activeCircle: Circle,
  circles: Circle[],
  shot: ShotData,
  walls: Rect[],
  onComplete: () => void
): void {
  const state: PhysicsState = {
    activeCircle,
    circles,
    shot,
    walls,
    step: 0,
    slowFrames: 0,
    onComplete
  };

  applyImpulse(state.activeCircle, state.shot);
  requestAnimationFrame(() => physicsStep(state));
}

function physicsStep(state: PhysicsState): void {
  integrate(state.circles);
  resolveCircleWallCollisions(state.circles, state.walls);
  resolveCircleCircle(state.circles);
  applyFriction(state.circles);

  if (allStopped(state.circles)) {
    state.slowFrames++;
    if (state.slowFrames >= settleNeed) {
      resetVelocities(state.circles);
      state.onComplete();
      return;
    }
  } else {
    state.slowFrames = 0;
  }

  state.step++;
  if (state.step < maxSteps) {
    requestAnimationFrame(() => physicsStep(state));
  } else {
    state.onComplete(); // force exit
  }
}